src/NUnitFramework/framework/Constraints/Comparers/ChainComparer.cs

```diff
@@ -0,0 +1,127 @@
+// ***********************************************************************
+// Copyright (c) 2018 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace NUnit.Framework.Constraints.Comparers
+{
+    /// <summary>
+    /// An <see cref="IEqualityComparer"/> that also supports equality comparisons
+    /// with a <see cref="Tolerance"/>.
+    /// </summary>
+    internal abstract class ChainComparer : IEqualityComparer
```

Comment:
Can you help me understand why we're implementing both non-generic and generic versions of `IEqualityComparer`?

Reply:
The non-generic `ChainComparer` is the default base class for special-case NUnit comparisons. The generic `ChainComparer<T>` is used as base class for special-case comparisons that involve comparing two values of the same type (such as `DateTime` and `string`) and reduces boiler-plate code for these comparisons.